File: app/lib/util/polyline.dart
import 'package:latlong2/latlong.dart';

/// Google encoded polyline(정밀도 1e-5)을 좌표 목록으로 푼다. OTP legGeometry.points 형식.
List<LatLng> decodePolyline(String encoded) {
  final points = <LatLng>[];
  var index = 0;
  var lat = 0;
  var lon = 0;
  while (index < encoded.length) {
    int shift = 0;
    int result = 0;
    int b;
    do {
      b = encoded.codeUnitAt(index++) - 63;
      result |= (b & 0x1f) << shift;
      shift += 5;
    } while (b >= 0x20);
    lat += (result & 1) != 0 ? ~(result >> 1) : (result >> 1);

    shift = 0;
    result = 0;
    do {
      b = encoded.codeUnitAt(index++) - 63;
      result |= (b & 0x1f) << shift;
      shift += 5;
    } while (b >= 0x20);
    lon += (result & 1) != 0 ? ~(result >> 1) : (result >> 1);

    points.add(LatLng(lat / 1e5, lon / 1e5));
  }
  return points;
}

File: backend/internal/crossing/polyline.go
// Package crossing 은 도보·따릉이 구간이 지나는 신호 횡단보도를 세고 기대 대기를 더한다.
// 입력은 otp/extract_crossings.py 가 서울 OSM 에서 뽑은 crossings.csv(신호 있는 횡단보도 노드 좌표).
package crossing

// Point 는 위경도(도).
type Point struct{ Lat, Lon float64 }

// DecodePolyline 은 Google encoded polyline(정밀도 1e-5, OTP legGeometry.points)을 점 목록으로 푼다.
// 형식이 깨지면 그때까지 푼 점만 돌려준다.
func DecodePolyline(s string) []Point {
	var out []Point
	lat, lon := 0, 0
	i := 0
	next := func() (int, bool) {
		result, shift := 0, 0
		for {
			if i >= len(s) {
				return 0, false
			}
			b := int(s[i]) - 63
			i++
			result |= (b & 0x1f) << shift
			shift += 5
			if b < 0x20 {
				break
			}
		}
		if result&1 != 0 {
			return ^(result >> 1), true
		}
		return result >> 1, true
	}
	for i < len(s) {
		dLat, ok := next()
		if !ok {
			break
		}
		dLon, ok := next()
		if !ok {
			break
		}
		lat += dLat
		lon += dLon
		out = append(out, Point{Lat: float64(lat) / 1e5, Lon: float64(lon) / 1e5})
	}
	return out
}
